Improve nargs

hokacci opened the pull request:
Hi.
Thank you for this nice library.

I see some issues regarding "remaining" behavior.

https://github.com/p-ranav/argparse/issues/118
https://github.com/p-ranav/argparse/issues/81

I also wanted optional argument preceded by variable length positional arguments (including zero length) like Python argparse nargs "*", "+", "?"

So, I implemented it just for experiment.
For now, this worked for me.

I know this change drops "remaining" feature, which would impact existing projects.
Also, this seems a complicated problem and I think more study may be needed.

What do you think about this change?

Thank you in advance.

hokacci commented:
I looked into this more precisely and revived remaining() partially.

The only difference is below.

https://github.com/p-ranav/argparse/pull/125/files#diff-dc3197c89fb7cbd8d7e504f8ac7ffc2cfbae3cc56a1c17e672ec0a96256b61f8L62

skrobinson commented:
The `std::numeric_limits` class needs the `limits` header. Otherwise, tests all pass.  Good job.

But, there's many changes here and I may have more questions.

hokacci commented:
I included "limits" and  added some test cases.

Thank you.

skrobinson commented:
Please rebase on current master. While you're rebasing, I'd like to see a reduced number of commits and expanded messages telling why the commit was added. Trying to do per commit reviews was counter-productive due to later commits undoing earlier changes.

Should SizeRange be renamed (e.g. NArgsRange) to indicate its specific purpose? I'm undecided on this and would like to hear from others.

Can the two std::distance checks be shortened? I ask because this looks like duplication to be refactored, but I don't see a simple answer. Maybe it is best as the checks currently work.

I see an opportunity to refactor your updated `Argument::validate` for DRY and a shorter function. The following example compiles, but fails tests all over the test suite. (And I'm not a fan of my six-condition `if`, either.)

```cpp
  void validate() const {
    std::stringstream stream;
    if (mIsOptional) {
      if ((mIsUsed && !mNumArgsRange.contains(mValues.size()) &&
           !mIsRepeatable && !mDefaultValue.has_value()) ||
          (!mNumArgsRange.contains(mValues.size()) && !mDefaultValue.has_value())) {
        if (!mUsedName.empty())
          stream << mUsedName << ": expected ";
        if (mNumArgsRange.is_exact()) {
          stream << mNumArgsRange.get_min();
        } else if (mNumArgsRange.is_right_bounded()) {
          stream << mNumArgsRange.get_min() << " to " << mNumArgsRange.get_max();
        } else {
          stream << mNumArgsRange.get_min() << " or more";
        }
        stream << " argument(s). " << mValues.size() << " provided.";
      } else {
        // TODO: check if an implicit value was programmed for this argument
        if (!mIsUsed && !mDefaultValue.has_value() && mIsRequired) {
          stream << mNames[0] << ": required.";
        }
        if (mIsUsed && mIsRequired && mValues.size() == 0) {
          stream << mUsedName << ": no value provided.";
        }
      }
    }
    if (stream.str().size() > 0)
      throw std::runtime_error(stream.str());
  }
```

The project's preferred format for constructors appears to be (see `Argument` ctor):

```cpp
  SizeRange(std::size_t aMin, std::size_t aMax)
      : mMin(aMin), mMax(aMax) {
    if (aMin > aMax)
      throw std::logic_error("Range of number of arguments is invalid");
  }
```

The added `details::is_container_v<T>` check in `Argument::get` strikes me as another DRY opportunity.

I've learned by reading your code. I'll see something that strikes me as odd. But research shows me why you made that decision and I generally agree with your choice.

hokacci commented:
Thank you for your feedback.

I rebased this on master.

I added some explanation below.
I hope this helps you.

https://github.com/hokacci/argparse/commit/c6c3be04d3533c8fb7b52ed04ad1ffe986da7a61
https://github.com/hokacci/argparse/commit/3d559d3a23cf8567b1f300a1e8e70b8d87f90f0c

And I dealt with ALL_CAPS issue and member initializer list preference.

https://github.com/hokacci/argparse/commit/bec93acaa774ba8d66502332e4c88a6c6838eb45
https://github.com/hokacci/argparse/commit/14abaa47d9b0660fd1729c9e0bc9b0903999ac7c

Later, I will look at other issues you pointed out.

Thank you.

skrobinson commented:
Thank you for rebasing and consolidating. And your extended commit comments helped me confirm what you intended to do matched the actual changes.

Why do you think it's "too much" to raise an exception in .get()? I'm not disagreeing, I just don't understand why you prefer the new behavior.

In `test/test_actions.hpp`:

> [](std::string &sum, std::string const &value) { sum += value; }

Is there a reason for `const` after, rather than before, the type?

No rush, but I'd like to see `Argument::validate` simplified before this branch is merged.

Thank you for the many test cases, they helped me better understand your changes.

EchoPouet commented:
Hello.
When will this feature be implemented ?

hokacci commented:
@EchoPouet 
Hi, thank you for reminding me of this PR.
I was so busy and did not have enough time for this PR.

@skrobinson 
> Why do you think it's "too much" to raise an exception in .get()? I'm not disagreeing, I just don't understand why you prefer the new behavior.
It seemed natural for me. But now I think compatibility counts. So I fixed it.

> Is there a reason for const after, rather than before, the type?
No. Seemingly this project uses pre-const style. Now I also took this style.

> No rush, but I'd like to see Argument::validate simplified before this branch is merged.
I wrote some helper functions and make validate() clearer.

Also, I merged current master branch and fixed many conflicts.

Thank you.

EchoPouet commented:
Thank you @hokacci. I'm looking forward to the merge and the release.

p-ranav commented on include/argparse/argparse.hpp:
```diff
@@ -327,6 +328,43 @@ template <class T> struct parse_number<T, chars_format::fixed> {
+};
+
+enum class NArgsPattern {
+  ZeroOrOne,
```
For consistency, let's use lower_case underscore-separated field names for `enum class`es.

hokacci replied:
Fixed in https://github.com/p-ranav/argparse/pull/125/commits/b869b5a209f6f2963988fbcd5a7ec62ddd60d573

p-ranav commented:
I'd also love for the README to be updated as part of the PR, to describe the new version of `.nargs(...)`.

skrobinson commented:
There have been many changes since you branched from the main line. Some of the current merge conflict fixes undo recent changes, e.g. commit 37a1f3b9e6 in PR #172 and commit 748bc95cf5 in PR #142. I'd like to see this feature added without preventable side-effects.

hokacci commented:
> I'd also love for the README to be updated as part of the PR, to describe the new version of `.nargs(...)`.

@p-ranav 
I added minimal explanation in https://github.com/p-ranav/argparse/pull/125/commits/e44023f424c16a0136238624222b2e9a0b33b9b6

hokacci commented:
@skrobinson 

> There have been many changes since you branched from the main line. Some of the current merge conflict fixes undo recent changes, e.g. commit [37a1f3b](https://github.com/p-ranav/argparse/commit/37a1f3b9e6ddb27ad70fb3b52c83266066949488) in PR #172 and commit [748bc95](https://github.com/p-ranav/argparse/commit/748bc95cf5881394680cdc91c8027984427a5e6f) in PR #142. I'd like to see this feature added without preventable side-effects.

For commit [37a1f3b](https://github.com/p-ranav/argparse/commit/37a1f3b9e6ddb27ad70fb3b52c83266066949488),  I retrieved it in https://github.com/p-ranav/argparse/pull/125/commits/5d6544a04e8bf47d739f7088c95e1f84cf149c16
It was my mistake. Sorry.

For commit [748bc95](https://github.com/p-ranav/argparse/commit/748bc95cf5881394680cdc91c8027984427a5e6f), it seems the change is already retrieved.

Thank you.

hokacci commented:
I also did some small fixes in the recent three commits.
nargs_pattern::zero_or_one -> nargs_pattern::optional
SizeRange -> NArgsRange
and moved NArgsRange to private in Argument class. Users don't need to use this class.
